src/components/Swap/SwapButton/index.tsx

```diff
@@ -67,80 +67,82 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
+        // submit will open the status dialog and immediately cover the animating button.
+        return new Promise((resolve) => {
+          const onAnimationEnd = () => {
+            document.removeEventListener('animationend', onAnimationEnd)
+            resolve(true)
+          }
+          document.addEventListener('animationend', onAnimationEnd)
+        })
+      }
+      return true
+    },
+    [addTransactionInfo, resetSwapAmount, setDisplayTxHash]
+  )
 
   const [isPending, setIsPending] = useState(false)
   const native = useNativeCurrency()
   const onWrap = useCallback(async () => {
     setIsPending(true)
-    try {
-      const transaction = await wrapCallback()
-      if (!transaction) return
-      invariant(wrapType !== undefined)
-      addTransaction({
-        response: transaction,
-        type: wrapType,
-        amount: CurrencyAmount.fromRawAmount(native, transaction.value?.toString() ?? '0'),
-      })
-      setDisplayTxHash(transaction.hash)
-      // Only reset pending after any queued animations to avoid layout thrashing, because a
-      // successful wrap will open the status dialog and immediately cover the button.
-      const postWrap = () => {
-        setIsPending(false)
-        document.removeEventListener('animationend', postWrap)
-      }
-      if (isAnimating(document)) {
-        document.addEventListener('animationend', postWrap)
-      } else {
-        setIsPending(false)
-      }
-    } catch (e) {
-      // TODO(zzmp): Surface errors from wrap.
-      console.log(e)
-    } finally {
-      setIsPending(false)
-    }
-  }, [addTransaction, native, setDisplayTxHash, wrapCallback, wrapType])
+    await onSubmit(async () => {
+      const response = await wrapCallback()
+      if (!response) return
+
+      invariant(wrapType)
+      const amount = CurrencyAmount.fromRawAmount(native, response.value?.toString() ?? '0')
+      return { response, type: wrapType, amount }
+    })
+
+    // Whether or not the transaction submits, we should still reset the pending state.
+    setIsPending(false)
+  }, [native, onSubmit, wrapCallback, wrapType])
   // Reset the pending state if user updates the swap.
   useEffect(() => setIsPending(false), [inputCurrencyAmount, trade])
 
+  const setOldestValidBlock = useSetOldestValidBlock()
   const onSwap = useCallback(async () => {
-    try {
-      const transaction = await swapCallback?.()
-      if (!transaction) return
-      invariant(trade.trade)
-      addTransaction({
-        type: TransactionType.SWAP,
-        response: transaction,
-        tradeType: trade.trade.tradeType,
-        trade: trade.trade,
-      })
-      setDisplayTxHash(transaction.hash)
+    const submitted = await onSubmit(async () => {
+      const response = await swapCallback?.()
+      if (!response) return
 
       // Set the block containing the response to the oldest valid block to ensure that the
```

Comment:
shouldn't all these `oldest` references be `newest`?

Reply:
we want the `oldest` valid block because that implies that nothing older than that is still valid.
I'm happy to change this to something less ambiguous in a separate PR, but would prefer not to change the naming in this PR.